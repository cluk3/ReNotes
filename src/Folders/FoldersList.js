import React, { PureComponent } from "react";
import Folder from "./Folder";
import NewFolder from "./NewFolder";
import NewFolderInput from "./NewFolderInput";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import { createNewFolder, setActiveFolder } from "./stateManager";
import { getDefaultValue } from "../helpers";
import styled from "styled-components";
import { ENTITIES } from "../constants";
import PropTypes from "prop-types";

const FoldersUl = styled.ul`
  margin: 0;
  padding: 0;
`;

const FolderListContainer = styled.div`
  height: 100%;
  overflow-y: auto;
  border-right: #dedede solid 1px;
`;

export class FoldersList extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      creationMode: false
    };
  }

  static propTypes = {
    folders: PropTypes.array.isRequired,
    activeFolder: PropTypes.string.isRequired,
    isFolderFocused: PropTypes.bool.isRequired,
    createNewFolder: PropTypes.func.isRequired,
    setActiveFolder: PropTypes.func.isRequired
  };

  componentWillUpdate(nextProps) {
    const { folders, activeFolder } = this.props;
    if (folders.length > nextProps.folders.length) {
      const folderToDeleteIndex = folders.indexOf(activeFolder);
      let newActiveFolder;
      if (folders.length === 1) {
        // it was the last in the array
        newActiveFolder = "";
      } else if (folderToDeleteIndex === folders.length - 1) {
        // it was in the last position
        newActiveFolder = folders[folderToDeleteIndex - 1];
      } else {
        newActiveFolder = folders[folderToDeleteIndex + 1];
      }
      this.props.setActiveFolder(newActiveFolder);
    }
  }

  handleFolderClick(activeFolderName) {
    this.props.activeFolder !== activeFolderName &&
      this.props.setActiveFolder(activeFolderName);
  }

  handleNewFolderClick() {
    this.setState({
      creationMode: true
    });
  }

  handleSubmit(newFolderName) {
    const success = true;
    if (this.props.folders.some(folderName => folderName === newFolderName)) {
      return !success;
    } else {
      this.setState({
        creationMode: false
      });
      this.props.createNewFolder(newFolderName);
      return success;
    }
  }

  render() {
    const { isFolderFocused, activeFolder, folders } = this.props;
    const foldersList = this.props.folders.map(folderName => (
      <Folder
        name={folderName}
        key={folderName}
        handleFolderClick={id => this.handleFolderClick(id)}
        selected={activeFolder === folderName}
        highlighted={activeFolder === folderName && isFolderFocused}
      />
    ));

    return (
      <FolderListContainer>
        <FoldersUl>{foldersList}</FoldersUl>
        {this.state.creationMode && (
          <NewFolderInput
            handleSubmit={newFolderName => this.handleSubmit(newFolderName)}
            defaultValue={getDefaultValue(folders)}
          />
        )}
        <NewFolder handleClick={() => this.handleNewFolderClick()} />
      </FolderListContainer>
    );
  }
}

function mapStateToProps({ folders, focusedElement }) {
  return {
    folders: folders.allNames,
    activeFolder: folders.activeFolder,
    isFolderFocused: focusedElement.elementType === ENTITIES.FOLDERS
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      createNewFolder,
      setActiveFolder
    },
    dispatch
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(FoldersList);
