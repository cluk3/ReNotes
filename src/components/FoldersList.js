import React, { Component } from "react";
import Folder from "./Folder";
import NewFolder from "./NewFolder";
import _ from "lodash";

class FoldersList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      folders: []
    };
  }

  handleFolderClick() {}

  handleNewFolderClick() {
    this.setState(prevState => ({
      folders: [...prevState.folders, { creationMode: true, id: _.uniqueId() }]
    }));
  }

  handleSubmit(name, event) {
    event.preventDefault();
    this.setState(prevState => ({
      folders: [
        ...prevState.folders.slice(0, -1),
        Object.assign({}, prevState.folders[prevState.folders.length - 1], {
          name,
          creationMode: false
        })
      ]
    }));
  }

  render() {
    const folders = this.state.folders.map(({ name, id, creationMode }) => (
      <Folder
        name={name}
        key={id}
        handleClick={() => this.handleFolderClick()}
        creationMode={creationMode}
        handleSubmit={(name, event) => this.handleSubmit(name, event)}
      />
    ));

    return (
      <div>
        <ul>{folders}</ul>
        <NewFolder handleClick={() => this.handleNewFolderClick()} />
      </div>
    );
  }
}

export default FoldersList;
