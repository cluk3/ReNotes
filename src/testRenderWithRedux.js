import React from 'react';
import { createStore } from 'redux';
import { Provider, connect } from 'react-redux';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { initialState as notesInitialState } from './Notes/modules/notes';
import { initialState as foldersInitialState } from './Folders/modules/folders';
import { initialState as focusedElementInitialState } from './containers/Focusable/stateManager';

const reducer = state => state;
const defaultInitialState = {
  folders: foldersInitialState,
  notes: notesInitialState,
  focusedElement: focusedElementInitialState
};

export function renderWithRedux(
  ui,
  {
    initialState = defaultInitialState,
    store = createStore(reducer, initialState)
  } = {}
) {
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    // adding `store` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    store
  };
}
