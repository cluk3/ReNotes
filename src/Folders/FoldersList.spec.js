import React from 'react';
import { renderWithRedux } from '../testRenderWithRedux';
import FoldersList from './FoldersList';
import 'jest-styled-components';

test('it works', () => {
  const { container } = renderWithRedux(<FoldersList />);
  expect(container.firstChild).toMatchSnapshot();
});
