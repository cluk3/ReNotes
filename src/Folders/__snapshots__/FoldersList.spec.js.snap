// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoldersList renders an input for modifying a folder when it is in edit mode 1`] = `
.c5 {
  list-style: none;
  padding: 0.4em 0 0.4em 1em;
  background-color: #f3f4f5;
}

.c7 {
  fill: #888889;
}

.c6 {
  position: absolute;
  bottom: 0;
  border: none;
  font-size: 16px;
  background-color: #f3f4f5;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.5em 0.5em;
}

.c8 {
  padding-left: 0.3em;
}

.c3 {
  outline: none;
  width: 99%;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  font: 500 16px sans-serif;
  border: none;
  padding: 0;
}

.c2 {
  position: relative;
  z-index: 110;
  background-color: #4286f4;
  padding: 0.4em 0 0.4em 1em;
}

.c4 {
  display: block;
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  z-index: 100;
}

.c1 {
  margin: 0;
  padding: 0;
}

.c0 {
  height: 100%;
  overflow-y: auto;
  border-right: #dedede solid 1px;
}

<div
  class="c0"
>
  <div
    class="react-contextmenu-wrapper"
  >
    <ul
      class="c1"
    >
      <div>
        <form
          class="c2"
        >
          <input
            aria-label="Edit folder name"
            class="c3"
            type="text"
            value="Test Folder"
          />
        </form>
        <div
          class="c4"
        />
      </div>
      <li
        class="c5"
      >
        Test Folder 2
      </li>
    </ul>
  </div>
  <button
    aria-label="Add New Folder"
    class="c6"
  >
    <svg
      class="c7"
      height="20px"
      viewBox="0 0 30 30"
      width="20px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M15,3C8.373,3,3,8.373,3,15c0,6.627,5.373,12,12,12s12-5.373,12-12C27,8.373,21.627,3,15,3z M21,16h-5v5 c0,0.553-0.448,1-1,1s-1-0.447-1-1v-5H9c-0.552,0-1-0.447-1-1s0.448-1,1-1h5V9c0-0.553,0.448-1,1-1s1,0.447,1,1v5h5 c0.552,0,1,0.447,1,1S21.552,16,21,16z"
      />
    </svg>
    <span
      class="c8"
    >
      New Folder
    </span>
  </button>
  <nav
    class="react-contextmenu"
    role="menu"
    style="position: fixed; opacity: 0; pointer-events: none;"
    tabindex="-1"
  >
    <div
      aria-disabled="false"
      class="react-contextmenu-item"
      role="menuitem"
      tabindex="-1"
    >
      Rename Folder...
    </div>
    <div
      aria-disabled="false"
      class="react-contextmenu-item"
      role="menuitem"
      tabindex="-1"
    >
      Delete Folder...
    </div>
    <div
      aria-disabled="false"
      aria-orientation="horizontal"
      class="react-contextmenu-item react-contextmenu-item--divider"
      role="menuitem"
      tabindex="-1"
    />
    <div
      aria-disabled="false"
      class="react-contextmenu-item"
      role="menuitem"
      tabindex="-1"
    >
      New Folder...
    </div>
  </nav>
</div>
`;

exports[`FoldersList works 1`] = `
.c2 {
  list-style: none;
  padding: 0.4em 0 0.4em 1em;
  background-color: #3b90fe;
}

.c3 {
  list-style: none;
  padding: 0.4em 0 0.4em 1em;
  background-color: #f3f4f5;
}

.c5 {
  fill: #888889;
}

.c4 {
  position: absolute;
  bottom: 0;
  border: none;
  font-size: 16px;
  background-color: #f3f4f5;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.5em 0.5em;
}

.c6 {
  padding-left: 0.3em;
}

.c1 {
  margin: 0;
  padding: 0;
}

.c0 {
  height: 100%;
  overflow-y: auto;
  border-right: #dedede solid 1px;
}

<div
  class="c0"
>
  <div
    class="react-contextmenu-wrapper"
  >
    <ul
      class="c1"
    >
      <li
        class="c2"
      >
        Test Folder
      </li>
      <li
        class="c3"
      >
        Test Folder 2
      </li>
    </ul>
  </div>
  <button
    aria-label="Add New Folder"
    class="c4"
  >
    <svg
      class="c5"
      height="20px"
      viewBox="0 0 30 30"
      width="20px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M15,3C8.373,3,3,8.373,3,15c0,6.627,5.373,12,12,12s12-5.373,12-12C27,8.373,21.627,3,15,3z M21,16h-5v5 c0,0.553-0.448,1-1,1s-1-0.447-1-1v-5H9c-0.552,0-1-0.447-1-1s0.448-1,1-1h5V9c0-0.553,0.448-1,1-1s1,0.447,1,1v5h5 c0.552,0,1,0.447,1,1S21.552,16,21,16z"
      />
    </svg>
    <span
      class="c6"
    >
      New Folder
    </span>
  </button>
  <nav
    class="react-contextmenu"
    role="menu"
    style="position: fixed; opacity: 0; pointer-events: none;"
    tabindex="-1"
  >
    <div
      aria-disabled="false"
      class="react-contextmenu-item"
      role="menuitem"
      tabindex="-1"
    >
      Rename Folder...
    </div>
    <div
      aria-disabled="false"
      class="react-contextmenu-item"
      role="menuitem"
      tabindex="-1"
    >
      Delete Folder...
    </div>
    <div
      aria-disabled="false"
      aria-orientation="horizontal"
      class="react-contextmenu-item react-contextmenu-item--divider"
      role="menuitem"
      tabindex="-1"
    />
    <div
      aria-disabled="false"
      class="react-contextmenu-item"
      role="menuitem"
      tabindex="-1"
    >
      New Folder...
    </div>
  </nav>
</div>
`;
